src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.HandleFileMove.cs

```diff
@@ -0,0 +1,95 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Runtime.Remoting.Contexts;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.VisualStudio.ProjectSystem.Waiting;
+using Microsoft.VisualStudio.Settings;
+using Solution = Microsoft.CodeAnalysis.Solution;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Rename
+{
+    internal partial class RenamerProjectTreeActionHandler
+    {
+        private readonly HashSet<Renamer.RenameDocumentActionSet> _results = new HashSet<Renamer.RenameDocumentActionSet>();
+
+        protected virtual async Task CPSCopyAsync(IProjectTreeActionHandlerContext context, IReadOnlyList<IProjectTree> items, DropEffects effect)
+        {
+            await base.CopyAsync(context, items, effect);
+        }
+
+        public override async Task CopyAsync(IProjectTreeActionHandlerContext context, IReadOnlyList<IProjectTree> items, DropEffects effect)
+        {
+            Requires.NotNull(context, nameof(Context));
+            Requires.NotNull(items, nameof(items));
+
+            _results.Clear();
+
+            ISettingsManager settings = await _settingsManagerService.GetValueAsync();
+            bool EnableNamespaceUpdate = settings.GetValueOrDefault("SolutionNavigator.EnableNamespaceUpdate", false);
+
+            if (EnableNamespaceUpdate)
+            {
+                CodeAnalysis.Project? project = GetCurrentProject();
+
+                if (project is null)
+                {
+                    return;
+                }
+
+                foreach (var item in items)
+                {
+                    string oldFilePath = ""; // Get filename
+                    string filenameWithExtension = ""; // remote extension
+                    string destinationFolder = ""; // Get destination folder name;
+
+                    (bool result, Renamer.RenameDocumentActionSet? documentActions)
+                        = await GetRenameSymbolsActions(project, oldFilePath, filenameWithExtension, new List<string>(new string[] { destinationFolder }));
+
+                    if (result == true)
+                    {
+                        _results.Add(documentActions!);
+                    }
+                }
+            }
+
+            await CPSCopyAsync(context, items, effect);
+
+            foreach (var documentAction in _results)
+            {
+
+                _threadingService.RunAndForget(async () =>
+                {
+                    Solution currentSolution = await PublishLatestSolutionAsync();
+
+                    await _projectVsServices.ThreadingService.SwitchToUIThread();
+
+                    string updateNamespaceAction = documentAction.ApplicableActions.First().GetDescription(CultureInfo.CurrentCulture);
```

Comment:
Does `documentAction.ApplicableActions` always contain at least one item? If not, this will throw.